### ISSUE TITLE
NoExpectedFormOption when logging into any other account

### ISSUE THREAD

**Issue by surfdaworld** [costastf/locationsharinglib#41]
I have a successful authentication and cookie file from one user account on my gsuite domain. I wanted to switch to a different user for location tracking, but I am unable to authenticate for any other user on my domain using get-maps-cookies.  I have verified that the non-working accounts do not have 2FA enabled.

Unsure how to troubleshoot from here, as all accounts are set up the same way. 1 works, the other 2 do not.

Just a note regarding usage, I am trying to get this working for my local installation of HomeAssistant.

Exact command being used:
`sudo get-maps-cookies -e USERNAME -p PASSWORD -c cookies.cookies`

**Comment by costastf** [comment]
You don't need sudo to execute this and you should not really use sudo for things that do not really require it. Please run with debug like 
```get-maps-cookies -L DEBUG -e USERNAME -p PASSWORD -c cookies.cookies``` and after sanitizing the output of any personal info please paste it here so i can try to figure out what is happening.

**Comment by eirinnm** [comment]
I'm new to this library - trying it for the first time today - but I've got the same bug. Here's the full traceback:

```
Traceback (most recent call last):
  File "c:\users\eirinn\anaconda3\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\users\eirinn\anaconda3\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\Eirinn\Anaconda3\Scripts\get-maps-cookies.exe\__main__.py", line 9, in <module>
  File "c:\users\eirinn\anaconda3\lib\site-packages\locationsharinglib\cli.py", line 137, in main
    CookieGetter(args.email, args.password, args.cookies_file)
  File "c:\users\eirinn\anaconda3\lib\site-packages\locationsharinglib\locationsharinglib.py", line 331, in __init__
    cookies_file=cookies_file)
  File "c:\users\eirinn\anaconda3\lib\site-packages\locationsharinglib\locationsharinglib.py", line 214, in __init__
    self._authenticate()
  File "c:\users\eirinn\anaconda3\lib\site-packages\locationsharinglib\locationsharinglib.py", line 340, in _authenticate
    selection_form = self._skip_challenge(response)
  File "c:\users\eirinn\anaconda3\lib\site-packages\locationsharinglib\locationsharinglib.py", line 299, in _skip_challenge
    form = self._get_required_form(response, '/signin/challenge/skip')
  File "c:\users\eirinn\anaconda3\lib\site-packages\locationsharinglib\locationsharinglib.py", line 288, in _get_required_form
    raise NoExpectedFormOption
locationsharinglib.locationsharinglibexceptions.NoExpectedFormOption
```

**Comment by costastf** [comment]
This means that the account is not setup properly for this to work. You need to have push authentication enabled on your account. Please look into the issue tracker for discussions about seeing up the account properly.

**Comment by eirinnm** [comment]
Oh I see. I already had 2FA set up with Google Authenticator but I've now enabled the push authenticator service too. However the same error occurs. Do I need to disable the Google Authenticator?

**Comment by jamime** [comment]
@zouden you need to disable 2FA and keep push authentication according to #27 
@costastf I'm also seeing the error
```
(googlemaps-r3ImGlvk) bash-3.2$ get-maps-cookies -e xxx@gmail.com -p xxx -L DEBUG
Traceback (most recent call last):
  File "/Users/jamime/.local/share/virtualenvs/googlemaps-r3ImGlvk/bin/get-maps-cookies", line 11, in <module>
    sys.exit(main())
  File "/Users/jamime/.local/share/virtualenvs/googlemaps-r3ImGlvk/lib/python3.7/site-packages/locationsharinglib/cli.py", line 137, in main
    CookieGetter(args.email, args.password, args.cookies_file)
  File "/Users/jamime/.local/share/virtualenvs/googlemaps-r3ImGlvk/lib/python3.7/site-packages/locationsharinglib/locationsharinglib.py", line 331, in __init__
    cookies_file=cookies_file)
  File "/Users/jamime/.local/share/virtualenvs/googlemaps-r3ImGlvk/lib/python3.7/site-packages/locationsharinglib/locationsharinglib.py", line 214, in __init__
    self._authenticate()
  File "/Users/jamime/.local/share/virtualenvs/googlemaps-r3ImGlvk/lib/python3.7/site-packages/locationsharinglib/locationsharinglib.py", line 340, in _authenticate
    selection_form = self._skip_challenge(response)
  File "/Users/jamime/.local/share/virtualenvs/googlemaps-r3ImGlvk/lib/python3.7/site-packages/locationsharinglib/locationsharinglib.py", line 299, in _skip_challenge
    form = self._get_required_form(response, '/signin/challenge/skip')
  File "/Users/jamime/.local/share/virtualenvs/googlemaps-r3ImGlvk/lib/python3.7/site-packages/locationsharinglib/locationsharinglib.py", line 288, in _get_required_form
    raise NoExpectedFormOption
locationsharinglib.locationsharinglibexceptions.NoExpectedFormOption
```

`-L DEBUG` doesn't appear to work either.

```
{
    "_meta": {
        "hash": {
            "sha256": "6e9105b042a934d6bc3b55927e483978f8296263c8b43242ba63ef67fc952629"
        },
        "pipfile-spec": 6,
        "requires": {
            "python_version": "3.7"
        },
        "sources": [
            {
                "name": "pypi",
                "url": "https://pypi.org/simple",
                "verify_ssl": true
            }
        ]
    },
    "default": {
        "beautifulsoup4": {
            "hashes": [
                "sha256:194ec62a25438adcb3fdb06378b26559eda1ea8a747367d34c33cef9c7f48d57",
                "sha256:90f8e61121d6ae58362ce3bed8cd997efb00c914eae0ff3d363c32f9a9822d10",
                "sha256:f0abd31228055d698bb392a826528ea08ebb9959e6bea17c606fd9c9009db938"
            ],
            "version": "==4.6.3"
        },
        "cachetools": {
            "hashes": [
                "sha256:4319bbb78172e7bcf99423e1ecd6914b32336ccfe97d2058ffe62e641a7f3abe",
                "sha256:ede01f2d3cbd6ddc9e35e16c2b0ce011d8bb70ce0dbaf282f5b4df24b213bc5d"
            ],
            "version": "==2.0.1"
        },
        "certifi": {
            "hashes": [
                "sha256:376690d6f16d32f9d1fe8932551d80b23e9d393a8578c5633a2ed39a64861638",
                "sha256:456048c7e371c089d0a77a5212fb37a2c2dce1e24146e3b7e0261736aaeaa22a"
            ],
            "version": "==2018.8.24"
        },
        "chardet": {
            "hashes": [
                "sha256:84ab92ed1c4d4f16916e05906b6b75a6c0fb5db821cc65e70cbd64a3e2a5eaae",
                "sha256:fc323ffcaeaed0e0a02bf4d117757b98aed530d9ed4531e3e15460124c106691"
            ],
            "version": "==3.0.4"
        },
        "idna": {
            "hashes": [
                "sha256:156a6814fb5ac1fc6850fb002e0852d56c0c8d2531923a51032d1b70760e186e",
                "sha256:684a38a6f903c1d71d6d5fac066b58d7768af4de2b832e426ec79c30daa94a16"
            ],
            "version": "==2.7"
        },
        "locationsharinglib": {
            "hashes": [
                "sha256:549031c5b0ea07ff64d9e237fe4f1bd606a51ce3384cde762fc88873fed32e81"
            ],
            "index": "pypi",
            "version": "==3.0.5"
        },
        "requests": {
            "hashes": [
                "sha256:63b52e3c866428a224f97cab011de738c36aec0185aa91cfacd418b5d58911d1",
                "sha256:ec22d826a36ed72a7358ff3fe56cbd4ba69dd7a6718ffd450ff0e9df7a47ce6a"
            ],
            "markers": "python_version != '3.2.*' and python_version >= '2.6' and python_version != '3.3.*' and python_version != '3.1.*' and python_version != '3.0.*' and python_version < '4'",
            "version": "==2.19.1"
        },
        "urllib3": {
            "hashes": [
                "sha256:a68ac5e15e76e7e5dd2b8f94007233e01effe3e50e8daddf69acfd81cb686baf",
                "sha256:b5725a0bd4ba422ab0e66e89e030c806576753ea3ee08554382c14e685d117b5"
            ],
            "markers": "python_version != '3.2.*' and python_version >= '2.6' and python_version != '3.3.*' and python_version != '3.1.*' and python_version != '3.0.*' and python_version < '4'",
            "version": "==1.23"
        }
    },
    "develop": {}
}
```

Tried with push authentication enabled and push authentication disabled (no 2FA). Any ideas?

**Comment by costastf** [comment]
What this all means is the you are not presented with the option to skip the 2fa challenge. That means that either your account is not set up correctly or that due to your specific location you are getting a response that google only uses in your location and is not accounted for. I have to look into why debugging does not work so you can use it and we can try to figure out what is wrong and how to fix it.

**Comment by jamime** [comment]
@costastf I also tried my account with regular username and password authentication without push with the same error.

Do you have any docs on how the account should be set up?

**Comment by costastf** [comment]
No, not really, and i think that there are differences depending on where you are located. You can read #27 and get an idea. Please also try v.3.0.6 where i have appended the response to the exception raised so we can debug easier. Please make sure that you prune any personal info from the response before you post back here.

**Comment by eirinnm** [comment]
With 3.0.6 I can redirect the error to an html file and view it. It's a sign-in page asking for the password and capcha.

**Comment by costastf** [comment]
Well, there you go... Not much the library can do about it. I guess you should double check your account settings and give it some time to settle between tries. Maybe rebooting your modem and getting a new ip would help?

**Comment by costastf** [comment]
Would you mind paying the output so I can make the error a bit more explicit?

**Comment by eirinnm** [comment]
How does the library normally bypass the need for a capcha? 

Here's the output, with personal details redacted: https://pastebin.com/Yyyd6M7X

**Comment by costastf** [comment]
It does not and it cannot. Essentially if you try to connect once it should just work if your account settings are correct. If you try multiple times and have failures then I guess the captcha kicks in. Not much that can be done I am afraid without turning this simple and small library into an ML beast.

**Comment by jamime** [comment]
I also get a captcha.

I was comparing to https://github.com/t4qjXH8N/ioBroker.google-sharedlocations/blob/master/lib/google_auth.js which has as a few additional headers. It also fails, at the moment but that's because it doesn't have a user-agent. I think locationsharinglib may be getting blocked based on the user agent.

**Comment by costastf** [comment]
You can try updating the agent in line 201 and see how that works for you. I think that Google does some identification based on the agent and the ip since the connection is identified as unique on the connected list. If changing the agent works for you it would be an idea to implement  randomly choosing from a pool of agents per connection.

**Comment by jamime** [comment]
@costastf no such luck. It would always be a cat and mouse game, google doesn't want automated logins. It's worth adding a ⭐️ on https://issuetracker.google.com/issues/62938530 to make it an official API.

**Comment by surfdaworld** [comment]
Sorry for the long absence, and thanks for the responses!  I tried the steps outlined, and I am also stuck at the captcha page.  I had to manually update locationsharinglib to the latest version in order to obtain the debug output, as the version installed by HomeAssistant did not give debug output.
I'll try playing around with the user agent and see if I can make any headway with that, since I'm able to log in from a browser on the same machine without it prompting for a captcha.

**Comment by surfdaworld** [comment]
No luck so far with changing the user agent.  Two potential solutions:
1. Use an app password--I can't currently get this to work, but theoretically isn't this exactly what app passwords are for? Outlook obviously can't navigate a captcha when logging into my gmail account with an app password.

2. Log into the account using a browser, and capture the cookie from the browser cache.

Seems like option 2 wouldn't be too difficult, will play around with it...

**Comment by costastf** [comment]
This should be fixed with v3.0.11.
